File: day01/part2.go
package main

import (
  "fmt"
  "strings"
  "strconv"
  "os"
)

func main() {
  content, err := os.ReadFile("input")
  if err != nil {
    panic(err)
  }

  lines := strings.Split(string(content), "\n")

  numberTokens := []string{
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
  }

  count := 0
  for _, line := range lines {
    var left, right byte

    if len(line) == 0 {
      continue
    }

    subloop:
    for i := 0; i < len(line); i++ {
      elem := line[i]
      if elem >= '0' && elem <= '9' {
        left = elem
        break subloop
      }

      for n, token := range numberTokens {
        if i+len(token) >= len(line) {
          continue
        }

        if line[i:i+len(token)] == token {
          left = byte(n+1) + '0'
          break subloop
        }
      }
    }

    subloop2:
    for i := len(line) - 1; i >= 0; i-- {
      elem := line[i]
      if elem >= '0' && elem <= '9' {
        right = elem
        break subloop2
      }

      for n, token := range numberTokens {
        if i-len(token) < 0 {
          continue
        }

        if line[i-len(token)+1:i+1] == token {
          right = byte(n+1) + '0'
          break subloop2
        }
      }
    }

    n, err := strconv.Atoi(string([]byte{left, right}))
    if err != nil {
      panic(err)
    }

    count += n
  }

  fmt.Printf("%d\n", count)
}

File: day01/part2.js
const fs = require("fs")

const numbersString = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

const numbersMap = {
  "one": 1,
  "two": 2,
  "three": 3,
  "four": 4,
  "five": 5,
  "six": 6,
  "seven": 7,
  "eight": 8,
  "nine": 9
}

const lines = fs.readFileSync("input", "utf-8").split("\n")

const values = lines.map(line => {
  line = line.split("")

  const numbers = line.map((value, index) => {
    if (!Number.isNaN(Number(value))) {
      return value
    }
    
    for (const s of numbersString) {
      if (line.join("").slice(index).startsWith(s)) {
        return String(numbersMap[s])
      }
    }

    return ""
  }).filter(Boolean)

  const first = numbers.find(x => !Number.isNaN(Number(x)))
  const last = numbers.findLast(x => !Number.isNaN(Number(x)))

  return Number(first + last)

}).filter(Boolean)

const sum = values.reduce((acc, number) => acc + number, 0)

console.log(sum)
